### PR TITLE
Two Stream and General Radiation Refactoring

### PR DESCRIPTION
**Pull request by rgknox**
### Description:

Synchronize with: https://github.com/ESCOMP/CTSM/pull/2265

This set of changes adds two-stream radiation to FATES, following methods of Sellers, Liou, Bonan, Longo et al. and others. This implementation uses a generalized solution to the two-stream equations, which allows for its use in various combinations of serial and parallel arranged scattering elements. The general solution is similar to that found in ED2. 
With this implementation comes various refactors to the radiation call sequence and code that works for both two-stream and norman.  

Notes:
1) When this set of changes is no-longer a draft, the expectation is that answers should be the same, however, various radiation specific diagnostics may have been fixed, added or removed
2) The groundwork for turning off high-frequency and multi-dimensional history output is being laid, and will be finalized in a later PR.
3) Two-stream and norman can be switched in the new radiation/FatesRadiationMemMod.F90 module. It is an integer constant.  This switch will be added to the fates parameter file at a later date
4) Two-stream does not require any new parameters, and uses the same parameters as Norman
5) There currently is no consolidation of cohorts into scattering media, in other words, every single cohort is an independent scattering element with the two-stream method. A consolidation method is forthcoming and may or may not be introduced in this set of changes. FWIW: without consolidation, the difference in timing is fairly similar (which is pretty cool!)
6) The two-stream solution uses LAPACK for matrix inversion, and has been designed to minimize memory usage, and minimize unecessary calls to calculate canopy scattering coeficients
7) The two-stream methods are encapsulated in the module [radiation/TwoStreamMLPEMod.F90](https://github.com/rgknox/fates/blob/two-stream-clean/radiation/TwoStreamMLPEMod.F90).  This module has no dependencies, although it still requires a way to pass in error logging and messaging.
8) Functional units tests have been implemented for the two-stream solution, and tests model predictions for generic canopy assemblages and generic light conditions
9) The methods to get the canopy parameters and intermediate scattering coeficients (such as scattering coefficient omega, backscatter fraction and optical depth), as well as the sun-shade fractions,  follow Gordon Bonan's methods, as detailed in the CLM technical manual, his multi-layer canopy code, and forthcoming manuscript
10) The matrix inversion for two-stream uses a hard-coded local array stack (200x200) for the solution. I want to change this to have a dynamicaly allocated scratch space, allocated for each thread (note: not for each site, cohort or patch, just one scratch array for each thread). This is because gfortran warned that this array is too large for some compilers.
11) Has code in place to address with minor additional changes: https://github.com/NGEET/fates/issues/1025

Link to documentation: https://www.overleaf.com/read/nwmhkyrvszqr

Compatible CTSM branch: https://github.com/rgknox/ctsm/tree/fates-twostream

### Collaborators:
@ckoven @mpaiao @gbonan @adrifoster @rosiealice @glemieux @JessicaNeedham Jennifer Kowalczyk

### Expectation of Answer Changes:

Ultimately, model predictions will be the same, with only differences in targeted radiation diagnostics

### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the in-code documentation .AND. (the [technical note](https://github.com/NGEET/fates-docs) .OR. the wiki) accordingly.
- [ ] I have read the [**CONTRIBUTING**](https://github.com/NGEET/fates/blob/master/CONTRIBUTING.md) document.
- [ ] FATES PASS/FAIL regression tests were run
- [ ] If answers were expected to change, evaluation was performed and provided

### Test Results:

CTSM (or) E3SM (specify which) test hash-tag:

CTSM (or) E3SM (specify which) baseline hash-tag:

FATES baseline hash-tag:

Test Output:

### CODE REVIEW

**Comment by rgknox** [comment]
Here is a comparison between the Two-stream and Norman schemes.  Site is BCI, it uses inventory initialization, static-stand-structure mode (kind of like SP, but for a site and inventory init, no dynamics), and local meteorology, from 2002-2004:

https://drive.google.com/file/d/1U3yA62fneGpbSw2AwK4H6CidAHdGe1hW/view?usp=sharing

As you can see, the two methods give fairly similar results.

**Comment by rgknox** [comment]
To do:
1) proper error and endruns in the two-stream code
2) thread specific scratch space
3) ...

**Comment by rgknox** [comment]
A few updates and a few tests have been conducted recently. Firstly, the memory has been restructured so that scratch space for the two-stream solver is retained at the site level. Also, provisions have been made  to force a tight closure between the albedo sent to the host and the absorbed radiation in the canopy and soil. This reduces errors on the closure from 1e-6 W/m2 to pass at 1e-8 W/m2, and is implemented more to pass the strict energy closer tests on the clm/elm-side of the code. 

Some tests:
4x5 grid, nocomp/fixed-biogeog, two-stream vs norman: https://drive.google.com/file/d/17bKbCcZCq_yTwHQNESEV_zRi5qQm-5fN/view?usp=sharing
4x5 grid, nocom & fixed-biogeog, two-stream (forced closure) vs two-stream (no forced closure): https://drive.google.com/file/d/1i_NE4p5CZ5cWKq0sC6M5zdltqYhbRTVt/view?usp=sharing

File: radiation/TwoStreamMLPEMod.F90 [hunk 1]
```diff
@@ -0,0 +1,1716 @@
...
+       ! avmu = 1/0  (1 - 0.5/0 * ln(0.5/0.5) ) but the limit approaches 1
+       ! a value of 0.0001 does not break numerics and generates an avmu of nearly 1
+
+       if( abs(rad_params%xl(ft)) <0.0001) rad_params%xl(ft)=0.0001_r8
+       
+       ! There must be protections on xl to prevent div0 and other weirdness
+       rad_params%phi1(ft) = 0.5_r8 - 0.633_r8*rad_params%xl(ft) - 0.330_r8*rad_params%xl(ft)*rad_params%xl(ft)
+       rad_params%phi2(ft) = 0.877_r8 * (1._r8 - 2._r8*rad_params%phi1(ft)) !0 = horiz leaves, 1 - vert leaves.
+       
+       avmu = (1._r8/rad_params%phi2(ft))* &
```
Comment (mpaiao) [inline radiation/TwoStreamMLPEMod.F90:678]: I haven't fully mapped the order of operations for the 2-stream, but would it make more sense to define avmu as part of rad_params? It seems avmu is calculated in multiple locations (at least here and in `ZenithPrep`), and having it in a single location might be better to ensure it is always defined consistently.

**Review by mpaiao — COMMENTED** [review]
@rgknox Thanks for implementing the two-stream radiation and sorry for taking so long to review this! I don't have any major comment, mostly some minor suggestions. I did spot a few places where the FATES solver seems a bit different from ED2, so I also included some questions that are probably just for clarification.

File: radiation/TwoStreamMLPEMod.F90 [hunk 2]
```diff
@@ -0,0 +1,1685 @@
...
+  !                         3 = thermal (not used at the moment)
+  !
+
+  use shr_log_mod   , only: errMsg => shr_log_errMsg
+  use shr_sys_mod   , only: shr_sys_abort
+  
+  implicit none
+  private
+
+  integer, parameter :: r8 = selected_real_kind(12)
```
[the remark this answers is not in the document]
Comment (rgknox) [reply radiation/TwoStreamMLPEMod.F90:33]: Potentially, I wanted this module to be completely independent of FATES, in case it was adopted by the HLMs as well.  But, seeing how the HLMs would also have access to FatesConstantsMod, and that has no dependencies, it's probably fine to do this. I'll make the change.

File: radiation/TwoStreamMLPEMod.F90 [hunk 3]
```diff
@@ -0,0 +1,1685 @@
...
+      r_diff_dn = this%band(ib)%Rbeam_atm*( &
+           scelb%Ad*exp(-scelg%Kb*vai) + &
+           scelb%B1d*scelb%lambda1_beam*exp(scelb%a*vai) + &
+           scelb%B2d*scelb%lambda2_beam*exp(-scelb%a*vai)) + &
+           this%band(ib)%Rdiff_atm*( & 
+           scelb%B1d*scelb%lambda1_diff*exp(scelb%a*vai) + &
+           scelb%B2d*scelb%lambda2_diff*exp(-scelb%a*vai))
+
+      if(debug)then
+         if(r_diff_dn.ne.r_diff_dn)then
```
[the remark this answers is not in the document]
Comment (rgknox) [reply radiation/TwoStreamMLPEMod.F90:373]: thanks, that is indeed more intuitive

File: radiation/TwoStreamMLPEMod.F90 [hunk 4]
```diff
@@ -0,0 +1,1716 @@
...
+    integer,intent(in)    :: icol
+    integer,intent(in)    :: ib
+    real(r8)              :: r_diff_dn
+
+    ! Rdn = Ad e−(Kbv) + Re + λ1 B1d e^(av) + λ2 B2d e^(−av)
+
+    associate(scelb => this%band(ib)%scelb(ican,icol), &
+         scelg => this%scelg(ican,icol) )
+
+      r_diff_dn = this%band(ib)%Rbeam_atm*( &
```
[the remark this answers is not in the document]
Comment (rgknox) [reply radiation/TwoStreamMLPEMod.F90:364]: Yeah, we do it slightly different.  Like you say, instead of the total equals 1, the two components are solved for a unit irradiance of one. That way we can scale the resulting cofficients simply by the actual irradiance once it becomes known. 

I couldn't remember if ED2 had this requirement where its forced to perform a normalized solution, but I remmber it having the normalized solution nonetheless.  With ELM/CLM, the albedo must be retrieved at the beginning of the time-step sequence, so that the atmospheric model can do its calculations, and thus provide the surface irradiance.  And thus, we use the surface irradiance, later in the time-step, to drive energy balance and photosynthesis.  

I can give this whole thing some attention and explain what is going on with comments

File: radiation/TwoStreamMLPEMod.F90 (hunk 1 above)
Comment (rgknox) [reply radiation/TwoStreamMLPEMod.F90:678]: great idea, will do

File: main/FatesHistoryInterfaceMod.F90 [hunk 5]
```diff
@@ -4408,125 +4726,111 @@ subroutine update_history_hifrq(this,nc,nsites,sites,bc_in,dt_tstep)
...
-               hio_parsun_top_si_can     => this%hvars(ih_parsun_top_si_can)%r82d, &
-               hio_parsha_top_si_can     => this%hvars(ih_parsha_top_si_can)%r82d, &
-               hio_maint_resp_unreduced_si  => this%hvars(ih_maint_resp_unreduced_si)%r81d, &
-               hio_tveg   => this%hvars(ih_tveg_si)%r81d)
-
+    real(r8) :: dt_tstep_inv          ! Time step in frequency units (/s)
+    
+    if(.not.hio_include_hifr_multi) return
+    
+    associate( hio_ar_si_scpf                      => this%hvars(ih_ar_si_scpf)%r82d, &
```
[the remark this answers is not in the document]
Comment (rgknox) [reply main/FatesHistoryInterfaceMod.F90:4739]: haha, yeah, I'm almost more excited about this part, those multi-dimensional history fields being prepped every half-hour are sucking up time!

File: main/FatesHistoryInterfaceMod.F90 [hunk 6]
```diff
@@ -4707,108 +4931,136 @@ subroutine update_history_hifrq(this,nc,nsites,sites,bc_in,dt_tstep)
...
-                     hio_parprof_dif_si_cnlfpft(io_si,cnlfpft_indx) = hio_parprof_dif_si_cnlfpft(io_si,cnlfpft_indx) + &
-                          cpatch%parprof_pft_dif_z(ican,ipft,ileaf) * cpatch%area * AREA_INV
-                     !
-                     ! summarize across all PFTs
-                     hio_parsun_z_si_cnlf(io_si,cnlf_indx) = hio_parsun_z_si_cnlf(io_si,cnlf_indx) + &
-                          cpatch%ed_parsun_z(ican,ipft,ileaf) * cpatch%area * AREA_INV
-                     hio_parsha_z_si_cnlf(io_si,cnlf_indx) = hio_parsha_z_si_cnlf(io_si,cnlf_indx) + &
-                          cpatch%ed_parsha_z(ican,ipft,ileaf) * cpatch%area * AREA_INV
+            
+            if_zenith: if(cpatch%solar_zenith_flag) then
```
[the remark this answers is not in the document]
Comment (rgknox) [reply main/FatesHistoryInterfaceMod.F90:4942]: I'll take a close look through here

File: radiation/TwoStreamMLPEMod.F90 (hunk 4 above)
[the remark this answers is not in the document]
Comment (mpaiao) [reply radiation/TwoStreamMLPEMod.F90:364]: So just to confirm, you are calling the two stream twice, the first time assuming that all the above-canopy downward irradiance is direct, and a second time when all the above-canopy downward irradiance is assumed diffuse?

File: main/FatesHistoryInterfaceMod.F90 (hunk 6 above)
[the remark this answers is not in the document]
Comment (rgknox) [reply main/FatesHistoryInterfaceMod.F90:4942]: I see what you mean.

All history variables are initialized to zero as a default, via a call like this:

https://github.com/NGEET/fates/blob/3966cc85f3308a1653db108ef153edca9dcec012/main/FatesHistoryInterfaceMod.F90#L4794

But, If we want to set a value to be ignored from the temporal mean, then we need to explicitly override that zero with an ignore value. This is very often the case for variables with an extra dimension, when certain fields in that dimension just aren't present, like if we don't have any cohorts in a certain size class, we don't want to report zero, we want to report unknown.

Here is an example:

https://github.com/NGEET/fates/blob/3966cc85f3308a1653db108ef153edca9dcec012/main/FatesHistoryInterfaceMod.F90#L5042C68-L5042C86

So I see that in your example, we do have some variables that use an extra dimension, such as canopy layer. Instead of setting them all to zero if cosz<=0, we should make sure they get zero's when appropriate, and the ignore value when appropriate (such as if an index has no representatives).

[the remark this answers is not in the document]
Comment (rgknox) [reply main/FatesHistoryInterfaceMod.F90]: so I reworked the history diagnostics code in this section quite a bit. It no longer relies on the zenith angle for filtering calculations, and does include zeros when there is no atmospheric radiation. It also makes sure that the ignore value is used for indices of multiplexed arrays that are invalid (for example in a leaf layer that is not occupied by any vegetation).

File: radiation/TwoStreamMLPEMod.F90 (hunk 1 above)
Comment (rgknox) [reply radiation/TwoStreamMLPEMod.F90]: done

File: biogeophys/FatesPlantRespPhotosynthMod.F90 [hunk 7]
```diff
@@ -399,504 +413,608 @@ subroutine FatesPlantRespPhotosynthDrive (nsites, sites,bc_in,bc_out,dtime)
...
+                                     par_per_sunla = (rd_abs_leaf*fsun + rb_abs_leaf) / (fsun*cohort_layer_elai(iv))
+                                  else
+                                     par_per_sunla = 0._r8
+                                  end if
+                                  par_per_shala = rd_abs_leaf*(1._r8-fsun) / ((1._r8 - fsun)*cohort_layer_elai(iv))
+                                  
+                               else
+                                  
+                                  par_per_sunla = 0._r8
+                                  par_per_shala = 0._r8
```
[the remark this answers is not in the document]
Comment (rgknox) [reply biogeophys/FatesPlantRespPhotosynthMod.F90:675]: I could run some tests to see if the atmospheric boundary conditions for diffuse are ever non-zero for cos(zen) = 0.  If this does happen, we could consider doing something here, or in a new PR.  I suppose we use the zenith angle as the filter for performing the solve, because this happens before the downwelling diffuse boundary conditions is known, via a normalized calculation. 
Moving forward, for twilight, we could decide to solve if either of two conditions is true, if the cos(zen)>0, or if the previous atmospheric diffuse>0.  Aside from wasting a little time, there is no liability to running the solver and not using it if there is no diffuse light.  

If the diffuse scattering coefficients didn't change over the course of the day due to ice/wetness conditions, we could split the solver and just run the normalized diffuse calculations once per day when the canopy changes, and use that for all diffuse boundary conditions, but changing wetness/ice is a thing.

File: radiation/TwoStreamMLPEMod.F90 (hunk 4 above)
[the remark this answers is not in the document]
Comment (mpaiao) [reply radiation/TwoStreamMLPEMod.F90:366]: Just to confirm, this is needed because the host land models need direct and diffuse albedo, correct? I am asking because using a single scaling factor can be computationally more efficient, though it doesn't allow distinguishing direct and diffuse albedo.

File: biogeophys/FatesPlantRespPhotosynthMod.F90 (hunk 7 above)
[the remark this answers is not in the document]
Comment (mpaiao) [reply biogeophys/FatesPlantRespPhotosynthMod.F90:675]: I think this depends on the atmospheric boundary conditions, gridded ones may or may not have, whereas tower-based ones may some irradiance during twilight, at least if they are from high latitudes. In any case, I think this makes more sense for a future PR, because this is not exclusive of two-stream.

For reference, in ED2 we use two criteria to test whether or not to call the radiation driver: if zen < 89.9° or so, then it is daytime. If 89.9° ≤ zen < 96° (end of civil twilight) *and* total incoming irradiance is greater than 0.5 W m<sup>−2</sup>, then it is twilight (diffuse only), otherwise it assumes night and skips the radiation solver altogether.

File: radiation/TwoStreamMLPEMod.F90 (hunk 4 above)
[the remark this answers is not in the document]
Comment (rgknox) [reply radiation/TwoStreamMLPEMod.F90:366]: we do need to provide diffuse and direct albedo

File: radiation/TwoStreamMLPEMod.F90 [hunk 8]
```diff
@@ -0,0 +1,1665 @@
...
+    real(r8) :: betab_veg  ! beam backscatter for vegetation (no snow)
+    real(r8) :: betab_om   ! multiplication of beam backscatter and reflectance
+    real(r8) :: om_veg     ! scattering coefficient for vegetation (no snow)
+
+    if( (cosz-1.0) > nearzero ) then
+       write(log_unit,*)"The cosine of the zenith angle cannot exceed 1"
+       write(log_unit,*)"cosz: ",cosz
+       write(log_unit,*)"TwoStreamMLPEMod.F90:ZenithPrep"
+       call endrun(msg=errMsg(sourcefile, __LINE__))
+    elseif(cosz<0._r8)then
```
[the remark this answers is not in the document]
Comment (rgknox) [reply radiation/TwoStreamMLPEMod.F90:872]: noting we continue this conversation elsewhere

File: radiation/TwoStreamMLPEMod.F90 (hunk 3 above)
[the remark this answers is not in the document]
Comment (rgknox) [reply radiation/TwoStreamMLPEMod.F90]: updated

**Comment by rgknox** [comment]
Been working on documentation, see here (work in progress):  https://www.overleaf.com/read/nwmhkyrvszqr

**Comment by rgknox** [comment]
I recently cleaned up the math in the solver a little bit, the solution is exactly the same as it was. The change was implemented specifically because instead of 4 combined terms, B1u, B1d, B2u, and B2d, we really only need a B1 and a B2 term.  See the documentation linked.
I ran tests to ensure that the results were exactly the same as previous, including a 120 year run at BCI and unit tests, and they were.

npp

fsni

File: main/FatesInventoryInitMod.F90 [hunk 9]
```diff
@@ -101,7 +101,7 @@ module FatesInventoryInitMod
                                                            ! defined in model memory and a physical
                                                            ! site listed in the file
 
-   logical, parameter :: do_inventory_out = .false.
+   logical, parameter :: do_inventory_out = .true.
```
[the remark this answers is not in the document]
Comment (rgknox) [reply main/FatesInventoryInitMod.F90:104]: yeah, this should not have been changed, reverting

File: main/EDPftvarcon.F90 [hunk 10]
```diff
@@ -1742,7 +1744,7 @@ subroutine FatesCheckParams(is_master)
 
      if(.not.is_master) return
 
-     if(radiation_model.ne.1) then
+     if(.not.any(radiation_model == [norman_solver,twostr_solver])) then
         write(fates_log(),*) 'The only available canopy radiation model'
         write(fates_log(),*) 'is the Norman scheme: fates_rad_model = 1'
         write(fates_log(),*) 'The two-stream scheme is not available yet'
```
[the remark this answers is not in the document]
Comment (adrifoster) [reply main/EDPftvarcon.F90:1797]: I can't leave a comment on this below, but on line 1923 the comment is incorrect (by bad!!) and should read:

`write(fates_log(),*) 'fates_rad_leaf_xl for pft ', ipft, ' is outside the allowed range of -0.4 to 0.6'`

can you add this? Sorry!!

File: main/EDTypesMod.F90 [hunk 11]
```diff
@@ -265,7 +265,24 @@ module EDTypesMod
                                             ! which is used for fixation
 
      
-     
+     ! Two-stream scratch arrays
+     real(r8), allocatable :: omega_2str(:,:)   ! This is the matrix that is inverted to solve
+                                                ! the linear system of equations in the two-stream
```
Comment (adrifoster) [inline main/EDTypesMod.F90:270]: I think eventually we want to get rid of `EDTypesMod`? Is there a better place for these matrices?

File: main/FatesRestartInterfaceMod.F90 [hunk 12]
```diff
@@ -3632,15 +3630,49 @@ subroutine update_3dpatch_radiation(this, nsites, sites, bc_out)
...
-                 call PatchNormanRadiation (currentPatch, &
-                      bc_out(s)%albd_parb(ifp,:), &
-                      bc_out(s)%albi_parb(ifp,:), &
-                      bc_out(s)%fabd_parb(ifp,:), &
-                      bc_out(s)%fabi_parb(ifp,:), &
-                      bc_out(s)%ftdd_parb(ifp,:), &
-                      bc_out(s)%ftid_parb(ifp,:), &
-                      bc_out(s)%ftii_parb(ifp,:))
-
+                 if_solver: if(radiation_model.eq.norman_solver) then
```
[the remark this answers is not in the document]
Comment (adrifoster) [reply main/FatesRestartInterfaceMod.F90:3658]: @mpaiao so the compiler will fail if you have multiple of the same variable in a `select case` block?

File: main/EDTypesMod.F90 (hunk 11 above)
Comment (rgknox) [reply main/EDTypesMod.F90:270]: putting them on the site structure is the best spot IMO.  At the patch level, we are using up a lot of memory. At the thread (proc) level, we end up creating large scratch spaces that aren't used (but maybe thats not a big deal) for many patches. So this can go whereever the site data structure goes, or wherever we dynamically allocate stuff at the thread level (but I don't think we do the later).

**Comment by mpaiao** [comment]
@adrifoster (somehow I couldn't find the reply option for  your comment on GitHub)

The compilation will fail if you have multiple cases pointing to the same value. For example, the following code will not compile:
```fortran
select case (x)
case (1)
   write (*,*) "x is 1"
case (2)
   write (*,*) "x is 2"
case (1)
   write (*,*) "is x 1 again?"
end select
```

But if you use `if`, the code will compile but the third case will never happen:

```fortran
if (x == 1) then
   write (*,*) "x is 1"
else if (x == 2) then
   write (*,*) "x is 2"
else if (x == 1) then
   write (*,*) "is x 1 again?"
end if
```

**Comment by adrifoster** [comment]
> @adrifoster (somehow I couldn't find the reply option for your comment on GitHub)
> 
> The compilation will fail if you have multiple cases pointing to the same value. For example, the following code will not compile:
> 
> ```fortran
> select case (x)
> case (1)
>    write (*,*) "x is 1"
> case (2)
>    write (*,*) "x is 2"
> case (1)
>    write (*,*) "is x 1 again?"
> end select
> ```
> 
> But if you use `if`, the code will compile but the third case will never happen:
> 
> ```fortran
> if (x == 1) then
>    write (*,*) "x is 1"
> else if (x == 2) then
>    write (*,*) "x is 2"
> else if (x == 1) then
>    write (*,*) "is x 1 again?"
> end if
> ```

oh that is pretty cool. thanks!

**Review by adrifoster — APPROVED** [review]

**Review by mpaiao — APPROVED** [review]

**Comment by rgknox** [comment]
Some updates from over the last couple of weeks:

I put some time into adding some protections on situations that caused div0s and near singularities.  These protections generated lower solver errors. I started investigating how low I could set the solver error tolerance, and remembered that we also use that tolerance to check and make sure that the areas of the scattering elements fill up the whole space of the patch.  

Therefore, I recently gave this section some attention:

https://github.com/NGEET/fates/blob/d35b9b1462f775e22d7fe608fb6f459effed1af3/radiation/FatesTwoStreamInterfaceMod.F90#L237-L253

That code makes sure that the combined area of any elements that share a layer in the patch exactly matches the footprint of the patch. This is important for us to make sure that solver errors are not contaminated by area effects.  The amount of area adjustments done to the elements is incredibly small.

File: biogeochem/EDCanopyStructureMod.F90 [hunk 13]
```diff
@@ -1759,28 +1690,29 @@ subroutine leaf_area_profile( currentSite )
...
+          currentPatch%canopy_mask(:,:) = 0
           do cl = 1,currentPatch%NCL_p
              do ft = 1,numpft
-                do  iv = 1, currentPatch%nrad(cl,ft)
+                do_leaflayer: do  iv = 1, currentPatch%nrad(cl,ft)
                    if(currentPatch%canopy_area_profile(cl,ft,iv) > 0._r8)then
                       currentPatch%canopy_mask(cl,ft) = 1
+                      exit do_leaflayer
+                   else
+                      exit do_leaflayer
```
[the remark this answers is not in the document]
Comment (mpaiao) [reply biogeochem/EDCanopyStructureMod.F90:1717]: @rgknox I see you have some comments on the need for the nrad loop. Does this mean that the duplication of `exit do_leaflayer` in both the `then` and `else` blocks is intentional? If so, please feel free to resolve this conversation.

[the remark this answers is not in the document]
Comment (rgknox) [reply biogeochem/EDCanopyStructureMod.F90:1717]: I think it's doing what its supposed to be doing, but its just coded up poorly. Maybe I forgot to finish it up.  Since the uppermost leaf layer is always present, if any of the layers have area, we only need to query the top index (ie iv==). I'll make a correction here. I'll remove the loop, just have it query leaf layer 1, and double check that the array is initialized to zero.

File: main/FatesInterfaceTypesMod.F90 [hunk 14]
```diff
@@ -245,7 +248,7 @@ module FatesInterfaceTypesMod
    integer , public, allocatable :: fates_hdim_levfuel(:)          ! fire fuel size class (fsc) dimension
    integer , public, allocatable :: fates_hdim_levcwdsc(:)         ! cwd class dimension
    integer , public, allocatable :: fates_hdim_levcan(:)           ! canopy-layer dimension 
-   integer , public, allocatable :: fates_hdim_levleaf(:)          ! leaf-layer dimension 
+   real(r8), public, allocatable :: fates_hdim_levleaf(:)          ! leaf-layer dimension, integrated VAI [m2/m2]
```
[the remark this answers is not in the document]
Comment (rgknox) [reply main/FatesInterfaceTypesMod.F90:267]: that's odd, that should still be integer, must be another self-sabotage, good catch @mpaiao

[the remark this answers is not in the document]
Comment (rgknox) [reply main/FatesInterfaceTypesMod.F90:267]: wait, strike that, that should indeed be a real array.  The unit on it is integrated leaf+stem area.  I'm seeing the hdim lev variables as a mix of real and integer.

**Comment by rgknox** [comment]
Closing as this is now superceded through #1141
